resources/aws.go

```diff
@@ -0,0 +1,334 @@
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	errwrap "github.com/pkg/errors"
+)
+
+func init() {
+	RegisterResource("aws", func() Res { return &AwsRes{} })
+}
+
+// AwsRes is an aws resource.
+type AwsRes struct {
+	BaseRes `yaml:",inline"`
+	State   string `yaml:"state"` // state: running, stopped, terminated
+	Region  string `yaml:"region"`
+	Type    string `yaml:"type"`
+	ImageID string `yaml:"imageid"`
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *AwsRes) Default() Res {
+	return &AwsRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *AwsRes) Validate() error {
+	if obj.State != "running" && obj.State != "stopped" && obj.State != "terminated" {
+		return fmt.Errorf("State must be 'running', 'stopped' or 'terminated'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *AwsRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsRes) Watch() error {
+	send := false
+	var exit *error
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	}))
+	svc := ec2.New(sess)
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.GetName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	awschan := make(chan string)
+	go func() {
+		for {
+			diOutput, err := svc.DescribeInstances(&diInput)
+			if err != nil {
+				errwrap.Wrapf(err, "error describing instances")
+			}
+			var waitInput ec2.DescribeInstancesInput
+			if len(diOutput.Reservations) == 1 {
+				waitInput = ec2.DescribeInstancesInput{
+					InstanceIds: []*string{diOutput.Reservations[0].Instances[0].InstanceId},
```

Comment:
Is there a guarantee that there will be at least one instances returned here? (otherwise this could panic)

Reply:
Yes, if there is no instance to return, then len(diOutput.Reservations) == 0.